docs: Module 3 Accounts, Transactions - code sample integration (B9lab updates 2-1-a) (#847)
cosmos/sdk-tutorials: b9lab-content/3-main-concepts/08-modules.md

* A boilerplate implementation of the Application Blockchain Interface (ABCI) that communicates with the underlying Tendermint consensus engine.
* A general-purpose data store that persists the module state called `multistore`.
* A server and interfaces to facilitate interactions with the node.

Modules implement the majority of application logic while **core** attends to wiring and infrastructure concerns and enables modules to be composed into higher-order modules.

A module defines a subset of the overall state using one or more key/value stores, known as `KVStore`, and a subset of message types that are needed by the application and do not exist yet. Modules also define interactions with other modules that do already exist.

For developers, most of the work involved in building a Cosmos SDK application revolves around building custom modules required by their application that do not exist yet, and integrating them with modules that do already exist into one coherent application. Existing modules can come either from the Cosmos SDK itself, or **come from third-party developers** and be downloaded from an online module repository.

## Module Components

> It is a best practice to define a module in the `x/moduleName` folder, not to be confused with the Cosmos SDK’s `x/` folder that already exists. For example, the module called Checkers would go in `x/Checkers`.

Reviewer: Already exists where, exactly? Let's try to be more clear on the distinction.

Reviewer 2: Updated.

> Modules implement several concerns:

Reviewer: ```suggestion
Modules implement several elements:
```

### Keeper

Keepers are the gatekeepers to the module’s store(s). It is mandatory to go through a module’s keeper in order to access the store(s). A keeper contains the layout of storage within the store and methods to update and inspect it. If you come from a Module-View-Controller (MVC) world, then it helps to think of the keeper as the controller.

> Other modules may need access to a store, but other modules are also potentially malicious or poorly written. For this reason, developers need to consider who/what should have access to their module store(s). Additionally, to prevent a module from accessing randomly another modules at runtime, a module that needs access to another needs to declare its intent to use another module at construction. At this point, such a module is granted a runtime key that lets it access the other module. Only modules that hold this key to a store can access the store.

Reviewer: ```suggestion
Other modules may need access to a store, but other modules are also potentially malicious or poorly written. For this reason, developers need to consider who/what should have access to their module store(s). Additionally, to prevent a module from randomly accessing another module at runtime, a module that needs access to another needs to declare its intent to use another module at construction. At this point, such a module is granted a runtime key that lets it access the other module. Only modules that hold this key to a store can access the store.
```

The only restrictions on error codes are the following:

* Must be greater than one, as a code value of one is reserved for internal errors.
* Must be unique within the module.

<HighlightBox type="info">

Note that the Cosmos SDK provides a core set of common errors. These errors are defined in `types/errors/errors.go`.

</HighlightBox>

### Wrapping

The custom module errors can be returned as their concrete type as they already fulfill the error interface. However, module errors can be wrapped to provide further context and meaning to failed execution.

Regardless if an error is wrapped or not, the Cosmos SDK's errors package provides an API to determine if an error is of a particular kind via `Is`.

### ABCI

If a module error is registered, the Cosmos SDK errors package allows ABCI information to be extracted through the `ABCIInfo` API. The package also provides `ResponseCheckTx` and `ResponseDeliverTx` as auxiliary APIs to automatically get `CheckTx` and `DeliverTx` responses from an error.

> <ExpansionPanel title="Show me some code for my checkers blockchain">

Reviewer: Another CTA:

```suggestion
## Next Up
Have a look at the example below or head straight to the [next section](./09-protobuf) to learn more about Protobuf!

<ExpansionPanel title="Show me some code for my checkers blockchain">
```

Reviewer 2: Updated.

Now your application is starting to take shape.

## The Checkers Module

When you create your checkers blockchain application, you can, and ought to, include a majority of the standard modules like `auth`, `bank`, and so on. With the Cosmos SDK boilerplate in place, the _checkers part_ of your checkers application will most likely reside in a single `checkers` module. This is the module that you author.

## Game Wager

Earlier the goal was to let players play with _money_. Here, with the introduction of modules like `bank`, you can start handling that.

The initial ideas are:

* When creating a game, the wager amount is declared.
* When doing their first move, which is interpreted as "challenge accepted", each player is billed the amount. Obviously the amount should not be deducted on the game creation, as it is good business to first ask for acceptance from the payer.
    * If the opponent rejects the game, or the game times out, at this point, then the first player gets refunded.
* Subsequent moves by a player do not cost anything.

> * If a game ends in a win or times out on a forfeit, then the winning player gets double the wager amount.

Reviewer: ```suggestion
* If a game ends in a win or times out on a forfeit, the winning player gets the total wager amount.
```
